### PR TITLE
Callback with url parsing error

### PR DESCRIPTION
**Pull request by raymondfeng**
### Description

We should report errors for url parsing.

#### Related issues

https://github.com/strongloop/loopback-next/issues/3248#issuecomment-506390270

- connect to <link_to_referenced_issue>

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: lib/mongodb.js [hunk 1]
```diff
@@ -304,29 +304,40 @@ MongoDB.prototype.connect = function(callback) {
       err,
       client
     ) {
-      if (!err) {
+      if (err) {
+        if (callback) callback(err);
```
Comment (jannyHou) [inline lib/mongodb.js]: I think `if (callback) callback(err);` should be after the debug log `if (self.debug) { }`

File: test/mongodb.test.js [hunk 2]
```diff
@@ -64,6 +64,19 @@ describe('lazyConnect', function() {
...
 
+  it('should report url parsing error', function(done) {
+    const ds = global.getDataSource({
+      url: 'mongodb://xyz:@127.0.0.1:4/xyz_dev_db',
+    });
+
+    ds.on('error', function(err) {
+      should.exist(err);
+      err.name.should.equal('MongoNetworkError');
+      err.message.should.match(/failed to connect to server/);
```
Comment (jannyHou) [inline test/mongodb.test.js]: would [this code](https://github.com/strongloop/loopback-connector-mongodb/pull/531/files#diff-567d120f151f51afa9605ebcf0f20106R329) include the url in the error message?

Comment (raymondfeng) [reply test/mongodb.test.js]: The url is included in the error by MongoDB driver.

File: lib/mongodb.js [hunk 3]
```diff
@@ -304,29 +304,40 @@ MongoDB.prototype.connect = function(callback) {
...
-          if (callback) callback(err, self.db);
-        });
-      } else {
-        if (self.debug || !callback) {
           g.error(
             '{{MongoDB}} connection is failed: %s %s',
             self.settings.url,
             err
           );
         }
```
Comment (bajtos) [inline lib/mongodb.js]: Have you considered extracting this long block of error-handling code into a shared function inside `connect`? 

```js
if (err) return onError(err);
self.client = client;
// The database name might be in the url
return urlParser(self.settings.url, self.settings, function(err, url) {
  if (err) onError(err);
  // success - carry on
}

function onError(err) {
  if (callback) callback(err);
  if (self.debug) {
    // ...
  }
}
```

Also is it a good idea to pass the error to callback before logging it? Wouldn't it produce confusing logs, where the caller prints its error information and he low-level details are printed only later? Personally, I would log the error first and call back second.

**Comment by bajtos** [comment]
@raymondfeng would you like to finish & land this pull request?

**Pull request merged**